ThreadPool.hpp

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright (c) 2012 Jakob Progsch, Václav Zeman
+
+This software is provided 'as-is', without any express or implied
+warranty. In no event will the authors be held liable for any damages
+arising from the use of this software.
+
+Permission is granted to anyone to use this software for any purpose,
+including commercial applications, and to alter it and redistribute it
+freely, subject to the following restrictions:
+
+   1. The origin of this software must not be misrepresented; you must not
+   claim that you wrote the original software. If you use this software
+   in a product, an acknowledgment in the product documentation would be
+   appreciated but is not required.
+
+   2. Altered source versions must be plainly marked as such, and must not be
+   misrepresented as being the original software.
+
+   3. This notice may not be removed or altered from any source
+   distribution.
+*/
+
+#ifndef THREAD_POOL_HPP
+#define THREAD_POOL_HPP
+
+// containers
+#include <vector>
+#include <queue>
+// threading
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <atomic>
+#include <future>
+// utility wrappers
+#include <memory>
+#include <functional>
+// exceptions
+#include <stdexcept>
+
+// std::thread pool for resources recycling
+class ThreadPool {
+public:
+    // the constructor just launches some amount of workers
+    ThreadPool(size_t threads_n = std::thread::hardware_concurrency(), const bool terminate = false) : stop(false), terminate(terminate)
+    {
+        if(!threads_n)
+            throw std::invalid_argument("more than zero threads expected");
+
+        this->workers.reserve(threads_n);
+        for(; threads_n; --threads_n)
+            this->workers.emplace_back(
+                [this]
+                {
+                    while(true)
+                    {
+                        std::function<void()> task;
+
+                        {
+                            std::unique_lock<std::mutex> lock(this->queue_mutex);
+                            if(!this->stop && this->tasks.empty())
```

Comment:
It seems to be a bit of a race condition. If I run it enough times (or lower 100000 to something like 100) it will eventually work. But I do wonder why it was changed to begin with, since the original code used the lambda version of condition.wait() (which under the hood uses a while loop):

    this->condition.wait(lock,
        [this]{ return this->stop || !this->tasks.empty(); });

http://en.cppreference.com/w/cpp/thread/condition_variable/wait